File: platforms/android/ClarityIME/app/src/main/java/com/coraz/clarityime/LocalApiAuth.kt
package com.coraz.clarityime

import android.content.Context
import java.io.File

/** Reads loopback API token for mutating core endpoints (v0.4+). */
object LocalApiAuth {
    private const val ENV_TOKEN = "CLARITYIME_API_TOKEN"
    private const val ENV_ROOT = "CLARITYIME_ROOT"

    fun token(context: Context? = null): String? {
        System.getenv(ENV_TOKEN)?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }
        context?.let {
            ClarityPrefs.prefs(it).getString(ClarityPrefs.KEY_LOCAL_API_TOKEN, null)
                ?.trim()
                ?.takeIf { t -> t.isNotEmpty() }
                ?.let { t -> return t }
        }
        tokenFileCandidates().firstNotNullOfOrNull { path ->
            runCatching {
                val f = File(path)
                if (f.isFile) f.readText().trim().takeIf { it.isNotEmpty() } else null
            }.getOrNull()
        }?.let { return it }
        return null
    }

    private fun tokenFileCandidates(): List<String> {
        val root = System.getenv(ENV_ROOT)?.trim()?.takeIf { it.isNotEmpty() }
        val paths = mutableListOf<String>()
        if (!root.isNullOrEmpty()) {
            paths.add("$root/data/.local_api_token")
        }
        // Termux sidecar default when user clones repo to ~/clarityime
        paths.add("/data/data/com.termux/files/home/clarityime/data/.local_api_token")
        return paths
    }
}

File: platforms/ios/Shared/LocalApiAuth.swift
import Foundation

/// Loopback API token for mutating ClarityIME core endpoints (v0.4+).
enum LocalApiAuth {
    private static let envToken = "CLARITYIME_API_TOKEN"
    private static let envRoot = "CLARITYIME_ROOT"
    private static let defaultsKey = "local_api_token"
    private static let appGroupKey = AppGroupConstants.localApiToken

    static func token() -> String? {
        if let env = ProcessInfo.processInfo.environment[envToken]?.trimmingCharacters(in: .whitespacesAndNewlines),
           !env.isEmpty {
            return env
        }
        if let suite = UserDefaults(suiteName: AppGroupConstants.suiteName),
           let stored = suite.string(forKey: appGroupKey)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
           !stored.isEmpty {
            return stored
        }
        if let stored = UserDefaults.standard.string(forKey: defaultsKey)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
           !stored.isEmpty {
            return stored
        }
        for path in tokenFileCandidates() {
            if let text = try? String(contentsOfFile: path, encoding: .utf8)
                .trimmingCharacters(in: .whitespacesAndNewlines),
               !text.isEmpty {
                return text
            }
        }
        return nil
    }

    private static func tokenFileCandidates() -> [String] {
        var paths: [String] = []
        if let root = ProcessInfo.processInfo.environment[envRoot]?
            .trimmingCharacters(in: .whitespacesAndNewlines),
           !root.isEmpty {
            paths.append((root as NSString).appendingPathComponent("data/.local_api_token"))
        }
        let home = NSHomeDirectory()
        paths.append((home as NSString).appendingPathComponent("clarityime/data/.local_api_token"))
        return paths
    }
}
